Add per-path queues, prioritize control messages over traffic

tewinget opened the pull request:
TODO: make sure that paths we create (as a client) are removed from the per-path queues.

TODO: research / trial & error on how best to manage messages sent per tick

notlesh commented on llarp/router/outbound_message_handler.cpp:
```diff
@@ -190,34 +216,165 @@ namespace llarp
+      PathID_t pathid = std::move(roundRobinOrder.front());
+      roundRobinOrder.pop();
+
+      if(outboundMessageQueues[pathid].size() > 0)
```
This will have the side effect of creating an empty entry in `outboundMessageQueues` when one doesn't exist (and, in these cases, this conditional will still evaluate to `false`). 

~~On that note, is there any logic to clean up the queues when they're no longer used (timeout, etc.)?~~

I see that these entries are erased below, but what about the side effect mentioned here?

tewinget replied:
if a `pathid` is in `roundRobinOrder` then it has a corresponding `outboundMessageQueues` entry.

notlesh commented on llarp/router/outbound_message_handler.hpp:
```diff
@@ -71,17 +89,41 @@ namespace llarp
     ILinkManager *_linkManager;
     std::shared_ptr< Logic > _logic;
+
+    const PathID_t zeroID;
```
A comment explaining this magic value might help people who aren't aware of its purpose

majestrate replied:
couldn't this be a static member?

tewinget replied:
Will make static and add comment.

tewinget replied:
Also, that reminds me that I need to make nodes reject paths with the zero ID, and path builds not use it.  Doing that now.

tewinget commented on llarp/net/net.cpp:
```diff
@@ -15,15 +15,17 @@
+#include <linux/if.h>
+extern "C" unsigned int
+if_nametoindex(const char* __ifname) __THROW;
+#endif
```
Must be -- I did `make format` before committing.

tewinget replied:
Also this is gone in dev now, so it's no longer relevant after I rebased.

tewinget commented on llarp/router/i_outbound_message_handler.hpp:
```diff
@@ -20,9 +20,14 @@ namespace llarp
 
+  static const size_t MAX_PATH_QUEUE_SIZE            = 40;
+  static const size_t MAX_OUTBOUND_QUEUE_SIZE        = 200;
+  static const size_t MAX_OUTBOUND_MESSAGES_PER_TICK = 20;
```
As far as I recall it's a semantic choice in this case but the two are equivalent.  Can change if desired though.

tewinget commented on llarp/router/outbound_message_handler.cpp:
```diff
@@ -190,34 +216,165 @@ namespace llarp
+    entry.message      = std::move(msg);
+    auto callback_copy = entry.message.second;
+    entry.router       = remote;
+    entry.pathid       = pathid;
```
True, but it also requires the caller to know that the value will be moved and act accordingly.  Tough call imo.  Will change if desired for certain.

notlesh replied:
I think the difference is that by taking a ref, you *always* require a copy, but by taking a value and doing a move, you sometimes incur one copy.

That said, this call is nothing compared to the network I/O that follows... :)

tewinget replied:
I agree that it would be slightly more cpu efficient to do a move, I just don't know if it's worth it to require any part of the code calling this function to be aware of that side-effect.  I can do it either way, of course.

tewinget commented on llarp/router/outbound_message_handler.cpp:
```diff
@@ -190,34 +216,165 @@ namespace llarp
+    else
+    {
+      removedSomePaths = false;
+    }
```
Yeah, that'd definitely better.  Not sure why I wrote it so verbosely.  Changing.

tewinget commented on llarp/router/outbound_message_handler.cpp:
```diff
@@ -190,34 +216,165 @@ namespace llarp
+
+  void
+  OutboundMessageHandler::SendRoundRobin()
+  {
```
This will only be called from the Logic thread, so it should not.  I could add a TRACEY lock that's null for release builds if desired?

tewinget commented:
> Is this still a WIP?

Apparently so, considering Travis doesn't seem to like it.  Taking a look here soon.

tewinget commented:
I derped and changed `OutboundMessageHandler::zeroID` to be static but didn't initialize/define it in the cpp file.  Fixed now, just double checking before I commit and push.

tewinget commented:
Ok, I think it's good to go now.  Paths build just fine with `make testnet` but of course it's harder to test whether path traffic is still fine.

majestrate commented on llarp/path/path.cpp:
```diff
@@ -31,8 +31,14 @@ namespace llarp
         hops[idx].rc = h[idx];
-        hops[idx].txID.Randomize();
-        hops[idx].rxID.Randomize();
+        while(hops[idx].txID.IsZero())
```
```
do
{
   randomize
}
while(iszero)
```

majestrate replied:
defensive style to prevent possible use of unzero'd but not random memory
